File: solutions/39/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Mon Nov  1 14:26:49 2021
> Description:   
 ************************************************************************/
#include <iostream> 
#include <vector> 
#include <numeric> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    vector<vector<int>> result;
    vector<int> path;
public:
    void backtracking(const vector<int>& candidates, const int target, const int startId) {
        // prune
        if (target < 0) return;
        // base case
        if (target == 0) {
            result.push_back(path);
            return;
        }
        for (int i = startId; i < candidates.size() && candidates[i] <= target; ++i) {
            path.push_back(candidates[i]);
            backtracking(candidates, target - candidates[i], i);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        backtracking(candidates, target, 0);
        return result;
    }
};

TEST(Test39, SimpleTest) {
    const int target = 7; 
    vector<int> candidates{
        2, 3, 6, 7
    };

    Solution s; 
    vector<vector<int>> result = s.combinationSum(candidates, target);
    for (auto r : result) {
        const int sum = accumulate(r.begin(), r.end(), decltype(r)::value_type(0));
        EXPECT_EQ(target, sum);
    }
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/39/main.py
class Solution:
    def __init__(self): 
        self.path = []
        self.result = []
    def backtracking(self, candidates, target, startId): 
        # prune 
        if (target < 0): 
            return 
        # base case 
        if (target == 0): 
            self.result.append(self.path[:])
            return 
        for i in range(startId, len(candidates)):
            if (candidates[i] > target):
                return 
            self.path.append(candidates[i])
            self.backtracking(candidates, target - candidates[i], i)
            self.path.pop()
            
    def combinationSum(self, candidates, target):
        candidates = sorted(candidates)
        self.backtracking(candidates, target, 0)
        return self.result

if __name__ == "__main__":
    candidates = [2, 3, 6, 7]
    target = 7
    s = Solution()
    print(s.combinationSum(candidates, target))
